### PR TITLE
feat(actions): add docs propagation action for columns

### PR DESCRIPTION
**Pull request by shirshanka**
This PR introduces a new DocPropagationAction for column-level documentation propagation. 

Key features include:

1. Automated propagation of documentation changes to downstream and sibling schema fields
2. Configuration options to enable/disable propagation at global and column levels
3. Support for ADD, MODIFY, and REMOVE operations on documentation
4. Integration with DataHub's event system, processing EntityChangeEvents and MetadataChangeLogs
5. Ability to fetch and apply global settings from DataHub
6. Detailed logging and statistics tracking for monitoring and debugging

As seen in DataHub townhall - https://youtu.be/hdqgjxiW-zI?t=2051

### CODE REVIEW

File: datahub-actions/src/datahub_actions/plugin/action/propagation/docs/propagation_action.py [hunk 1]
```diff
@@ -0,0 +1,645 @@
...
+        schema_field_urn: str,
+        dataset_urn: str,
+        field_doc: Optional[str],
+        context: SourceDetails,
+    ) -> Optional[MetadataChangeProposalWrapper]:
+        if context.origin == schema_field_urn:
+            # No need to propagate to self
+            return None
+
+        if not dataset_urn.startswith("urn:li:dataset"):
```
Comment (hsheth2) [inline datahub-actions/src/datahub_actions/plugin/action/propagation/docs/propagation_action.py:317]: we have utils e.g. `from datahub.metadata.urns import Dataset` or `guess_entity_type` - this startswith check feels a bit too hardcoded

File: datahub-actions/src/datahub_actions/plugin/action/propagation/docs/propagation_action.py [hunk 2]
```diff
@@ -0,0 +1,645 @@
...
+        )
+
+        source_details = context.dict(exclude_none=True)
+        attribution: MetadataAttributionClass = MetadataAttributionClass(
+            source=self.action_urn,
+            time=auditStamp.time,
+            actor=self.actor_urn,
+            sourceDetail=source_details,
+        )
+        documentations = graph.graph.get_aspect(schema_field_urn, DocumentationClass)
```
Comment (hsheth2) [inline datahub-actions/src/datahub_actions/plugin/action/propagation/docs/propagation_action.py:334]: not a huge fan of this `graph.graph.method` :( probably not worth fixing right now

File: datahub-actions/src/datahub_actions/plugin/action/propagation/docs/propagation_action.py [hunk 3]
```diff
@@ -0,0 +1,645 @@
...
+        return False
+
+    def get_upstreams(self, graph: AcrylDataHubGraph, entity_urn: str) -> List[str]:
+        """
+        Fetch the upstreams for an dataset or schema field.
+        Note that this DOES NOT support DataJob upstreams, or any intermediate nodes.
+        """
+        import urllib.parse
+
+        url_frag = f"/relationships?direction=OUTGOING&types=List(DownstreamOf)&urn={urllib.parse.quote(entity_urn)}"
```
Comment (hsheth2) [inline datahub-actions/src/datahub_actions/plugin/action/propagation/docs/propagation_action.py:451]: wasn't this method moved to the AcrylDataHubGraph? I think we can remove this

File: smoke-test/tests/consistency_utils.py
Comment (hsheth2) [inline smoke-test/tests/consistency_utils.py:1]: can we add a comment that says this was largely copied from the main datahub repo?

File: smoke-test/tests/actions/doc_propagation/test_propagation.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        if mcp.aspectName == "schemaMetadata":
+            dataset_urn = mcp.entityUrn
+            assert isinstance(mcp.aspect, models.SchemaMetadataClass)
+            schema_metadata: models.SchemaMetadataClass = mcp.aspect
+            for field in schema_metadata.fields:
+                field_urn = make_schema_field_urn(dataset_urn, field.fieldPath)
+                urns.append(field_urn)
+        return urns
+
+    from datahub.api.entities.dataset.dataset import Dataset
```
Comment (hsheth2) [inline smoke-test/tests/actions/doc_propagation/test_propagation.py:84]: why are these imports stuffed inside the methods? unless there's a good reason, they should be at the top of the file

**Review by hsheth2 — APPROVED** [review]

File: datahub-actions/src/datahub_actions/plugin/action/propagation/docs/propagation_action.py (hunk 1 above)
Comment (jjoyce0510) [reply datahub-actions/src/datahub_actions/plugin/action/propagation/docs/propagation_action.py:317]: i have seen this in a code quite a bit recently. i agree with avoiding raw strings for urn checks. seems like python sdk should have internal utilities for all of this

**Pull request merged**